File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/OpenAi.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.OpenAi;

/// <summary>
/// Structural client interface for OpenAI and OpenAI-compatible APIs
/// (Azure OpenAI, vLLM, TGI, LiteLLM proxy, Together, Groq, etc.).
/// </summary>
public interface IOpenAiClient
{
    OpenAiResponse ChatCompletionsCreate(JsonObject @params);
}

public sealed class Usage
{
    public long PromptTokens { get; set; }
    public long CompletionTokens { get; set; }
    public long TotalTokens { get; set; }
}

public sealed class ToolCall
{
    public string Id { get; }
    public ToolCall(string id) { Id = id; }
}

public sealed class Choice
{
    public IReadOnlyList<ToolCall> ToolCalls { get; }
    public bool HasFunctionCall { get; }
    public Choice(IReadOnlyList<ToolCall>? toolCalls, bool hasFunctionCall)
    {
        ToolCalls = toolCalls ?? new List<ToolCall>();
        HasFunctionCall = hasFunctionCall;
    }
}

public sealed class OpenAiResponse
{
    public string Model { get; }
    public Usage Usage { get; }
    public IReadOnlyList<Choice> Choices { get; }
    public OpenAiResponse(string model, Usage? usage, IReadOnlyList<Choice>? choices)
    {
        Model = model;
        Usage = usage ?? new Usage();
        Choices = choices ?? new List<Choice>();
    }
}

public sealed class OpenAiOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static OpenAiOptions Defaults() => new();
}

public sealed class BoundedOpenAiClient
{
    private readonly IOpenAiClient _raw;
    private readonly BoundedExecution _execution;
    private readonly OpenAiOptions _opts;

    public BoundedOpenAiClient(IOpenAiClient raw, BoundedExecution execution, OpenAiOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? OpenAiOptions.Defaults();
    }

    public OpenAiResponse ChatCompletionsCreate(JsonObject @params)
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();

        OpenAiResponse resp;
        try { resp = _raw.ChatCompletionsCreate(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        long tokens = resp.Usage.TotalTokens;
        if (tokens == 0) tokens = resp.Usage.PromptTokens + resp.Usage.CompletionTokens;
        long toolCalls = 0;
        foreach (var c in resp.Choices)
        {
            toolCalls += c.ToolCalls.Count;
            if (c.HasFunctionCall) toolCalls++;
        }
        string model = string.IsNullOrEmpty(resp.Model) ? "unknown" : resp.Model;
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"openai.chat.completions.create:{model}",
        });
        return resp;
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/openai.rs
//! OpenAI / OpenAI-compatible chat-completions wrapper. Satisfies
//! Azure OpenAI, vLLM, TGI, LiteLLM proxy, Together, Groq, etc.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Usage {
    #[serde(default)]
    pub prompt_tokens: u64,
    #[serde(default)]
    pub completion_tokens: u64,
    #[serde(default)]
    pub total_tokens: u64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ToolCall {
    pub id: String,
    #[serde(default)]
    pub r#type: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Message {
    #[serde(default)]
    pub tool_calls: Vec<ToolCall>,
    #[serde(default)]
    pub function_call: Option<Value>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Choice {
    #[serde(default)]
    pub message: Message,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Response {
    #[serde(default)]
    pub model: String,
    #[serde(default)]
    pub usage: Usage,
    #[serde(default)]
    pub choices: Vec<Choice>,
}

pub trait Client: Send + Sync {
    fn chat_completions_create(&self, params: &Value) -> Result<Response, String>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedClient<'a, C: Client> {
    raw: C,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, C: Client> BoundedClient<'a, C> {
    pub fn new(raw: C, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn chat_completions_create(&self, params: &Value) -> Result<Response, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self
            .raw
            .chat_completions_create(params)
            .map_err(MiddlewareError::Provider)?;
        let mut tokens = resp.usage.total_tokens;
        if tokens == 0 {
            tokens = resp.usage.prompt_tokens + resp.usage.completion_tokens;
        }
        let tool_calls = resp
            .choices
            .iter()
            .map(|c| c.message.tool_calls.len() as u64 + c.message.function_call.is_some() as u64)
            .sum();
        let model = if resp.model.is_empty() {
            "unknown".to_string()
        } else {
            resp.model.clone()
        };
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!("openai.chat.completions.create:{}", model)),
            ..Default::default()
        });
        Ok(resp)
    }
}
